clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -274,10 +273,10 @@ public ReadyCheckResult ready(Cluster cluster, long nowMs) {
             Deque<RecordBatch> deque = entry.getValue();
 
             Node leader = cluster.leaderFor(part);
-            if (leader == null) {
-                unknownLeadersExist = true;
-            } else if (!readyNodes.contains(leader)) {
-                synchronized (deque) {
+            synchronized (deque) {
+                if (leader == null && !deque.isEmpty()) {
```

Comment:
@rajinisivaram That removes elements from the deques held inside `this.batches`. But nothing ever removes entries from `this.batches` itself. My concern isn't so much the memory usage, rather that the # of entries in `this.batches` (even if the deque in the entry is empty) continues to grow. In the case of something like the REST proxy, this could grow to include the entire set of topic partitions, which will need to be iterated over for every single `RecordAccumulator.ready` call.